src/src-cpp/BjsonDecoder.hpp

```diff
@@ -0,0 +1,165 @@
+/*
+ * Copyright (c) 2017,2020 by Kemu Studio (visit ke.mu)
+ *
+ * Author(s): Sylwester Wysocki <sw@ke.mu>,
+ *            Roman Pietrzak <rp@ke.mu>
+ *
+ * This file is a part of the KEMU Binary JSON library.
+ * See http://bjson.org for more.
+ *
+ * Permission to use, copy, modify, and/or distribute this software for any
+ * purpose with or without fee is hereby granted, provided that the above
+ * copyright notice and this permission notice appear in all copies.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+ * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+ * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+ * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+ * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+ * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ */
+
+#include <string>
+#include <bjson/bjson-common.h>
+#include <bjson/bjson-decode.h>
+
+// -----------------------------------------------------------------------------
+//       Helper macros to wrap pure C calbacks into C++ member calls
+// -----------------------------------------------------------------------------
+
+#define BJSON_CPP_DECODE0(NAME)                              \
+  static int _handlerForPureC_##NAME(void *thiz)             \
+  {                                                          \
+    return (reinterpret_cast<BjsonDecoder*>(thiz))->NAME();  \
+  }
+
+#define BJSON_CPP_DECODE1(NAME, TYPE1)                           \
+  static int _handlerForPureC_##NAME(void *thiz, TYPE1 value)    \
+  {                                                              \
+    return (reinterpret_cast<BjsonDecoder*>(thiz))->NAME(value); \
+  }
+
+#define BJSON_CPP_DECODE2(NAME, TYPE1, TYPE2)                      \
+  static int _handlerForPureC_##NAME(void *thiz, TYPE1 x, TYPE2 y) \
+  {                                                                \
+    return (reinterpret_cast<BjsonDecoder*>(thiz))->NAME(x, y);    \
+  }
+
+
+class BjsonDecoder
+{
+  private:
```

Comment:
Methods below should be protected, right ?
```suggestion
  protected:
```